src/algorithms/xiao2013_utils.jl

```diff
@@ -0,0 +1,473 @@
+        end
+    end
+    rem_vertices!(g, vcat([twin_pair[1],twin_pair[2]],neighbor))
+
+    return true
+end
+        
+
+function first_twin(g::SimpleGraph)
+    vertices_with_degree_3 = [v for v in vertices(g) if degree(g, v) == 3]
+    for (v1, v2) in combinations(vertices_with_degree_3, 2)
+        (Set(neighbors(g, v1)) == Set(neighbors(g, v2))) && return (v1,v2)
+    end
+    return nothing
+end
+
+
+function is_complete_graph(g::SimpleGraph, vertices::Vector)
+    (length(vertices) <= 1) && return false
+    for (u, v) in combinations(vertices, 2)
+        (!has_edge(g, u, v)) && return false
+    end
+    return true
+end
+
+
+function short_funnel_filter!(g::SimpleGraph)
+    funnel_pair = first_short_funnel(g)
+    isnothing(funnel_pair) && return false
+
+    a,b = funnel_pair
+    N_a = neighbors(g, a)
+    N_b = neighbors(g, b)
+    for u in setdiff(N_a,vcat(N_b,[b])), v in setdiff(N_b,vcat(N_a,[a]))
+        (!has_edge(g,u,v)) && add_edge!(g,u,v)
+    end
+    rem_vertices!(g, [a,b])
+
+    return true
+end
+
+
+function first_short_funnel(g::SimpleGraph)
+    for a in vertices(g)
+        neighbors_a = neighbors(g, a)
+        for b in neighbors_a
+            neighbors_a_minus_b = setdiff(neighbors_a, [b])
+            if is_complete_graph(g, neighbors_a_minus_b)
+                neighbors_b_minus_a = setdiff(neighbors(g, b), [a])
+                num_nonadjacent_vertices = count(!has_edge(g, v1, v2) for v1 in neighbors_a_minus_b for v2 in neighbors_b_minus_a)
+                if length(intersect(neighbors(g, b),neighbors(g, a))) == 0 && num_nonadjacent_vertices <= degree(g, b)
+                    return (a,b)
+                end
+            end
+        end
+    end
+    return nothing
+end
+
+
+function desk_filter!(g::SimpleGraph)
+    desk_group = first_desk(g)
+    isnothing(desk_group) && return false
+
+    a,b,c,d = desk_group
+    for u in setdiff(open_neighbors(g, [a,c]), closed_neighbors(g, [b,d])) # N(a, c) - N[b, d]
+        for v in setdiff(open_neighbors(g, [b,d]), closed_neighbors(g, [a,c])) # N(b, d) - N[a, c]
+            !has_edge(g,u,v) && add_edge!(g,u,v)
+        end
+    end
+    rem_vertices!(g, [a,b,c,d])
+
+    return true
+end
+
+function first_desk(g::SimpleGraph)
+    all_cycles = simplecycles_limited_length(g, 4)
+    cycles_of_length_4 = filter(c -> (length(c) == 4 && c[end] > c[2]), all_cycles)
+    for quad in cycles_of_length_4
+        a, b, c, d = quad
+        if all(i -> degree(g, i) ≥ 3, quad)
+            if !(has_edge(g, c, a) || has_edge(g, d, b))
+                neighbor_A = open_neighbors(g, [a,c])
+                neighbor_B = open_neighbors(g, [b,d])
+                if isempty(intersect(neighbor_A,neighbor_B)) && length(setdiff(neighbor_A,[b,d])) <= 2 && length(setdiff(neighbor_B,[a,c])) <= 2
+                    return (a,b,c,d)
+                end
+            end
+        end
+    end
+    return nothing
+end
+
+function all_n_funnel(g::SimpleGraph, n::Int)

```

Comment:
use the same code as much as you can, here 3 and 4 funnel can be calculated by the same `n_funnel` function, then there is no need to write them separately.